Add equipment loadouts support for ModAccessorySlot

alueck opened the pull request:
### What is the new feature?
Equipment loadouts support for `ModAccessorySlot`
Also fixes #3285 

### Why should this be part of tModLoader?
Because modded accessory slots should support the vanilla equipment loadout switching

### Are there alternative designs?
The current implementation attempts to replicate the vanilla logic as closely as possible to facilitate adapting the code to future Terraria updates as easily as possible. 

### Sample usage for the new feature
https://github.com/tModLoader/tModLoader/blob/1.4.4/ExampleMod/Common/Players/ExampleAccessorySlot.cs shows the current examples of `ModAccessorySlot`. The `ExampleModWingSlot` example has been updated to support toggleable `HasEquipmentLoadoutSupport` in this PR: https://github.com/tModLoader/tModLoader/pull/4340/files#diff-f4ad9f01db3f16a5d9ef3f2e6aec684593990b7a861dc15fdee66056fe3ec2a6

Aside from setting `HasEquipmentLoadoutSupport` to an appropriate value, modders should not need to do anything different.  Action is only required if the modder wants the slot to not support loadouts. Modded accessories are still accessed in the same manner as before.

### Porting Notes
- `ModAccessorySlot` now supports equipment loadouts
  - By defaults loadout support will be true for any `ModAccessorySlot`. A modder can set the `ModAccessorySlot.HasEquipmentLoadoutSupport` property to `false` to disable it. The property value can also be toggled by a server-side config value, but a reload is required. When set to false, the `ModAccessorySlot` is "shared".
    - Changing the setting from `true` to `false` will cause the extra items stored to be spawned on the player when they enter the world.
    - Changing the value from `false` to `true` will result in the currently selected loadout holding the items.
    - `ModAccessorySlot` without custom backgrounds that support loadouts will be tinted with loadout colors as normal. If using a custom slot background, however, they will not be affected. A new hook, `ModAccessorySlot.BackgroundDrawColor`, can be used to further customize the color.
- `Player.DropItem` is now obsolete, use `Player.TryDroppingSingleItem` instead.
- `(Mod|Global)CanAccessoryBeEquippedWith/CanEquipAccessory` and `ModAccessorySlot.CanAcceptItem` are now potentially called on the server and remote clients instead of just the local client. This is to facilitate correctly syncing shared accessory slot conflicts. This means that there might be some syncing issues that need to be addressed.

### Discussion
- Support for additional loadouts is not included in this PR.
- `ModAccessorySlotPlayer.DropItems` has been changed to drop items even from disabled slots to match vanilla disabled slot logic. This change might affect some mod using slots for odd purposes.
- ~~I think `ModAccessorySlot.HasEquipmentLoadoutSupport` defaulting to `true` makes more sense. There was a brief discussion on this but I think we should revisit it.~~ We have gone with `true` now.
- This PR aligns drawing of modded slots with loadout support with vanilla drawing of accessory slots. The default background texture changed to `InventoryBack13`, which is basically white and is colored with a loadout and slot type specific color. Loadout colors are applied to `HasEquipmentLoadoutSupport` slots without a custom background texture.
  - Slots that do not support loadouts will continue drawing with the old textures. For players not using loadouts, they will be the same color, but players using loadouts will notice that "shared" slots stay the original green color to indicate that they are unchanged.

New background texture:
![Inventory_Back13](https://github.com/user-attachments/assets/5a84049c-cb92-4feb-bbcb-900a3c9073c0)
Old textures:
![Inventory_Back3](https://github.com/user-attachments/assets/35a08513-149d-4c15-b471-e0d39c68aef1)
![Inventory_Back8](https://github.com/user-attachments/assets/2e027380-627c-424b-9e17-8e55e19507c4)
![Inventory_Back12](https://github.com/user-attachments/assets/bcfaa850-9432-4907-b2bb-e0c86894408e)
Shared and non-shared ModAccessorySlot, and vanilla slots in loadout 1:
![image](https://github.com/user-attachments/assets/5bad3f4e-bb15-4574-9dca-a198ad8d204c)
The same in loadout 3:
![image](https://github.com/user-attachments/assets/9f836f7a-71c4-4eca-921e-dd9cf9804e5b)


### TODOs

- [x] Test multiplayer (tested locally with two tModLoader instances; synchronization worked fine)
- [x] Show info text if an item cannot be equipped in a shared slot, because of conflicts with any of the items equipped in any equipment loadout.
- [x] ~~Attempting to use a player that was saved while this PR was active prevents it from loading (OOB in ModAccessorySlotPlayer.LoadData). This will be an issue since we usually support loading players on older tModLoader versions, which users will be doing if anything goes wrong and they need to downgrade.~~ Fixed. But users will still lose accessories stored in ModAccessorySlot for the non-active loadouts if downgrading. This is fine.
- [x] Hook for customizing ModAccessorySlot draw color.

# Showcase
This video demonstrates the new behavior. There are several `ModAccessorySlot` active, 3 of them are shared. As the user switches loadouts, the shared slots stay green (like loadout 1) indicating that they are shared.

When the user switches to loadout 3, which has a functional wing accessory already, the shared wing accessory goes dark indicating that it is currently disabled. A chat message and hover message communicate to the player the issue. The player then moves the conflicting wing accessory and the slot becomes enabled again.

https://github.com/user-attachments/assets/74a8d906-7272-4a6a-9c41-7be9717e9aec

![image](https://github.com/user-attachments/assets/f6b348fc-c488-402a-9550-b539d297c7b4)

alueck committed:
Make ModAccessorySlots work with equipment loadouts

alueck committed:
Add support for mods which add additional loadouts

alueck committed:
Fix updating of dyes

alueck committed:
Add hook OnEquipmentLoadoutSwitched

alueck committed:
Fix UpdateDyes(bool) method

alueck committed:
Use actual number of vanilla loadouts for array intialization instead of assuming three loadouts

direwolf420 commented:
Over on the tmodloader discord there is a discussion ongoing about implementation details and use cases: https://discord.com/channels/103110554649894912/1148455439491682425 if you are interested

alueck committed:
Make more method internal

alueck committed:
Make ModAccessorySlot optionally shared between loadouts

alueck committed:
Fix CanSetItemForCurrentLoadout check

alueck committed:
Refactor code so only minimal changes on ItemSlot are needed

alueck committed:
Remove local function; use existing method instead

alueck committed:
Cleanup

alueck committed:
Remove line breaks

alueck committed:
Also load data to shared loadout in backwards compatibiliy case

alueck committed:
Use IsSharedSlot method instead of checking directly against the set

alueck committed:
Fix exception on data loading if there are unloaded slots

alueck committed:
Fix bug in loadout loading

alueck committed:
Fix data loading for freshyl migrated characters

alueck committed:
Move call to ResetAndSizeAccessoryArrays into LoadData

alueck committed:
Retrieve mod slot accessories outside loop in LoadPlayer_LastMinuteFixes

alueck committed:
Fix patch

alueck committed:
Disable loadout support by default

alueck committed:
Refactor code

alueck committed:
Fix condition

alueck committed:
Refactor data loading so no item will be lost if changing slot from shared to loadout support

alueck committed:
Merge branch '1.4.4' into mod-accessory-slot-loadouts-support

# Conflicts:
#	ExampleMod/Localization/TranslationsNeeded.txt

alueck committed:
Cleanup

alueck committed:
Change return type of ExEquipmentLoadout.LoadData

alueck committed:
Mark properties required

alueck committed:
Fix bug when loading data

Slots that belong to a different loadout only 100% changed their loadout support setting, if the slot contained at least one item in the save data

Solxanich committed:
Merge Conflicts Resolved -- PlayerLoader

JavidPack committed:
Merge branch '1.4.4' into pr/4340

JavidPack committed:
Fix some merge issues

JavidPack committed:
Restore UpdateVisibleAccessories wing slot logic

JavidPack committed:
Localization file updates

JavidPack committed:
Merge branch '1.4.4' into pr/4340

JavidPack committed:
Some patch fixes

JavidPack committed:
Update PlayerLoader.cs

JavidPack committed:
Remove unfinished additional loadout support, realign modslot approach more closely to vanilla logic.

JavidPack committed:
Cleanup

JavidPack committed:
ModAccessorySlot items now drop from player even if not usable (mediumcore, etc)

JavidPack committed:
Remove ExampleLoadoutSupportModAccessorySlot, move config example to ExampleModWingSlot

JavidPack committed:
Fix issue with "ghost" accessory item instances messing with post-mediumcore death accessory slots

JavidPack committed:
Obsolete Player.DropItem

JavidPack committed:
Fix unreported CanEquipAccessory issue when using over 10 modaccessoryslot

JavidPack committed:
Remove exAccessorySlotLoadoutConflict cache approach and just calculate on demand

JavidPack committed:
Fix AccCheck_ForPlayer code to use supplied player in called hooks.

JavidPack committed:
ModAccessorySlot.BackgroundDrawColor hook, custom textures no longer affected by loadout tint

JavidPack committed:
More todos looked into

JavidPack committed:
Verify more todos

JavidPack committed:
Update FNA

JavidPack committed:
Restore accidental localization entry deletion

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -325,55 +317,55 @@ public override void OnEquipmentLoadoutSwitched(int oldLoadoutIndex, int loadout
+		bool functionalIssue = Loader.IsAccessoryInConflict(Player, exAccessorySlot[functional], functional, Terraria.UI.ItemSlot.Context.ModdedAccessorySlot);
+		bool vanityIssue = Loader.IsAccessoryInConflict(Player, exAccessorySlot[vanity], vanity, Terraria.UI.ItemSlot.Context.ModdedVanityAccessorySlot);
+
+		return functionalIssue || vanityIssue;
```
It would be good to disable the vanity and functional slots independently. I'm not sure how feasible that is.

JavidPack replied:
I was able to implement this approach.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -503,9 +497,20 @@ public bool ModdedIsItemSlotUnlockedAndUsable(int index, Player player, bool ign
+		if (!canGoInSlot)
+			return true;
+		Item[] itemCollection = player.armor.Concat(ModSlotPlayer(player).exAccessorySlot).ToArray();
+		itemCollection[slot + player.armor.Length] = new Item();
```
Might be worth caching this `Item` instance

JavidPack committed:
Cache dummy item in check

JavidPack committed:
Disable functional and vanity slots due to conflicts individually.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -3,40 +3,65 @@
+	/// <summary> All slots that can potentially show for this user. First the loaded ModAccessorySlots, followed by any unloaded slots. </summary>
+	private readonly Dictionary<string, (int SlotType, bool HasLoadoutSupport)> slots = [];
+	/// <summary> Which slots are shared, indexed by ModAccessorySlot.Type </summary>
+	private readonly HashSet<int> sharedLoadoutSlotTypes = [];
```
Should still be an array.

JavidPack replied:
Done.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -3,40 +3,65 @@
 	internal bool[] exHideAccessory;
-	internal Dictionary<string, int> slots = new Dictionary<string, int>();
+	/// <summary> All slots that can potentially show for this user. First the loaded ModAccessorySlots, followed by any unloaded slots. </summary>
+	private readonly Dictionary<string, (int SlotType, bool HasLoadoutSupport)> slots = [];
```
Looks like this `bool HasLoadoutSupport` is unused now.

JavidPack replied:
Done.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -86,6 +126,26 @@ public override void LoadData(TagCompound tag)
+			return;
+
+		foreach (var items in extraItems) {
+			if(!items.Accessory.IsAir)
```
Might be worth running an `(if|for|foreach)\(` to find missing spaces periodically

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -193,16 +285,242 @@ public override void SendClientChanges(ModPlayer clientPlayer)
+			List<string> order,
+			Dictionary<string, (int SlotType, bool HasLoadoutSupport)> slots)
+		{
+			List<SlotInfo> result = [];
```
This should be `List<Item> itemsToDrop`. Modify the return type and `extraItems` accordingly

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -193,16 +285,242 @@ public override void SendClientChanges(ModPlayer clientPlayer)
+			Dictionary<string, (int SlotType, bool HasLoadoutSupport)> slots)
+		{
+			List<SlotInfo> result = [];
+			IList<Item> items;
+			IList<Item> dyes;
+			IList<bool> visible;
```
No need for forward declaration, just `var` below

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -193,16 +285,242 @@ public override void SendClientChanges(ModPlayer clientPlayer)
+
+				// If this slot has any acc/van/dye item, but doesn't have loadout support, the items shouldn't be in the loadout at all.
+				// We can try to move them to the inventory or try to put them in empty acc slots, but it might get complicated with slots and restrictions. Also the user might not want it there. They will be returned to the player OnEnterWorld.
+				bool hasLoadoutSupportSettingForSlotChanged = !hasLoadoutSupport && (!dye.IsAir || !accessory.IsAir || !vanityItem.IsAir);
```
Can simplify this with above ^. Just `if (!hasLoadoutSupport)` add each non-air item to `itemsToDrop`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -193,16 +285,242 @@ public override void SendClientChanges(ModPlayer clientPlayer)
+			for (int i = 0; i < order.Count; i++) {
+				(int type, bool hasLoadoutSupport) = slots[order[i]];
+
+				Item dye = dyes.ElementAtOrDefault(i) ?? new Item();
+				Item accessory = items.ElementAtOrDefault(i) ?? new Item();
+				Item vanityItem = items.ElementAtOrDefault(i + order.Count) ?? new Item();
+				bool isHidden = visible.ElementAtOrDefault(i);
```
Why are we using `ElementAtOrDefault`? Is there some way the data could be missing? Scary

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -193,16 +285,242 @@ public override void SendClientChanges(ModPlayer clientPlayer)
+		}
+	}
+
+	internal sealed record SlotInfo
```
Correspondingly, this class will be unused

JavidPack committed:
Handle no saved loadout tag more elegantly.

JavidPack committed:
Formatting

JavidPack committed:
Removed SlotInfo class

JavidPack committed:
Finish up feedback changes.

JavidPack committed:
HasEquipmentLoadoutSupport now defaults to true. Localization files updated
